cdap-app-fabric/src/main/java/co/cask/cdap/runtime/monitor/RuntimeMonitor.java

```diff
@@ -0,0 +1,177 @@
+    this.messagePublisher = messagePublisher;
+    this.clientConfig = clientConfig;
+    this.restClient = new RESTClient(clientConfig);
+    this.limit = cConf.getInt(Constants.RuntimeMonitor.RUNTIME_MONITOR_BATCH_LIMIT);
+    this.pollTimeMillis = cConf.getInt(Constants.RuntimeMonitor.RUNTIME_MONITOR_POLLTIME_MS);
+    this.topicsToRequest = new HashMap<>();
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    LOG.debug("Starting runtime monitor {}....", runtimeId);
+    initializeTopics();
+  }
+
+  private void initializeTopics() throws Exception {
+    Set<String> topicsToMonitor = new HashSet<>();
+    topicsToMonitor.addAll(Arrays.asList(cConf.get(Constants.RuntimeMonitor.RUNTIME_MONITOR_TOPICS).split(",")));
+
+    HttpResponse response = restClient.execute(HttpMethod.POST, clientConfig.resolveURL("runtime/monitor/topics"),
+                                               GSON.toJson(topicsToMonitor), Collections.emptyMap(), null);
+
+    if (response.getResponseCode() == HttpURLConnection.HTTP_UNAVAILABLE) {
+      throw new ServiceUnavailableException(response.getResponseBodyAsString());
+    }
+
+    Map<String, String> configToTopic = GSON.fromJson(response.getResponseBodyAsString(Charset.forName("UTF-8")),
+                                                      MAP_STRING_STRING_TYPE);
+
+    // TODO initialize from offset table for a given runtimeId
+    for (Map.Entry<String, String> entry : configToTopic.entrySet()) {
+      topicsToRequest.put(entry.getKey(), new MonitorConsumeRequest(entry.getValue(), entry.getKey(), null, limit));
+    }
+  }
+
+  @Override
+  protected void run() throws Exception {
+    runThread = Thread.currentThread();
+
+    try {
+      while (!stopped) {
+        try {
+          HttpResponse response = restClient.execute(HttpMethod.POST, clientConfig.resolveURL("runtime/metadata"),
+                                                     GSON.toJson(topicsToRequest), Collections.emptyMap(), null);
+
+          if (response.getResponseCode() == HttpURLConnection.HTTP_UNAVAILABLE) {
+            throw new ServiceUnavailableException(response.getResponseBodyAsString());
+          }
+
+          Map<String, List<MonitorMessage>> monitorResponses =
+            GSON.fromJson(response.getResponseBodyAsString(Charset.forName("UTF-8")), MAP_STRING_MESSAGE_TYPE);
+
+          processResponse(monitorResponses);
+        } catch (Exception e) {
+          OUTAGE_LOG.warn("Failed to fetch monitoring data from runtime {}. Will be retried in next iteration.",
+                          runtimeId, e);
+        }
+        Thread.sleep(pollTimeMillis);
+      }
+    } catch (InterruptedException e) {
+      // Interruption means stopping the service.
+    }
+  }
+
+  private void processResponse(Map<String, List<MonitorMessage>> monitorResponses) throws Exception {
+    for (String topicConfig : topicsToRequest.keySet()) {
+      MonitorConsumeRequest request = topicsToRequest.get(topicConfig);
```

Comment:
do the for loop on `topicsToRequest.entrySet()` so that you don't need to get from the map again.